plugins/manipulator.cpp

```diff
@@ -1253,150 +2309,396 @@ void viewscreen_unitlaborsst::refreshNames()
+
+    int bzone=display_rows/7;
+    int bottom_hint=(display_rows<=units.size()&&display_rows>16)?1:0;
+
+    if (first_row > units.size()-display_rows+bottom_hint ){
+        first_row = units.size()-display_rows+bottom_hint;
+    }
+
+    if (first_row < 0)
+        first_row = 0;
 
-    // if it shrinks vertically, scroll downward to keep the cursor visible
-    if (first_row < sel_row - num_rows + 1)
-        first_row = sel_row - num_rows + 1;
+    if(sel_row!=sel_row_b||display_rows!=display_rows_b)
+    {
+        if(first_row==0&&(sel_row_b<first_row||sel_row_b>first_row + display_rows)){
+            if(sel_row_b<sel_row){ //issued down
+                sel_row=sel_row_b; //just make sel row visible
+                first_row = sel_row +bzone+1 - display_rows;
+                row_hint=0;
+            }
 
-    // if the window grows horizontally, scroll to the left to eliminate blank columns from the right
-    if (first_column > NUM_COLUMNS - col_widths[DISP_COLUMN_LABORS])
-        first_column = NUM_COLUMNS - col_widths[DISP_COLUMN_LABORS];
+            if(sel_row_b>sel_row){ //issued up
+                sel_row=0;
+                sel_row_b=sel_row; //go to
+                row_hint=0;
+            }
+        }
+
+        //sel_row is pushing down
+        if( sel_row+bzone+1 > first_row + display_rows ){ //beyond zone
+            if(sel_row>sel_row_b)
+                first_row++;
+            if(sel_row+1+1 > first_row + display_rows){ //beyond max
+                first_row = sel_row + 1+1 - display_rows;
+                row_hint=0;
+            }
+        }
+
+        //sel_row is pushing up
+        if( sel_row-bzone < first_row ){
+            if(sel_row<sel_row_b)
+            first_row--;
+            if( sel_row-1 < first_row ){
+                first_row=sel_row-1;
+                row_hint=0;
+            }
+        }
+
+        if (first_row < 0){
+            first_row = 0;
+        }
+
+        sel_row_b=sel_row; display_rows_b=display_rows;
+    }
+
+    if (first_row > units.size()-display_rows+bottom_hint ){
+        first_row = units.size()-display_rows+bottom_hint;
+    }
+
+    if (first_row < 0)
+        first_row = 0;
+
+    int row_width=column_size[COLUMN_LABORS];
+    bzone=row_width/8+1;
+
+    //set first column according to if sel is pushing zone
+    if( sel_column-bzone < first_column )
+        first_column=sel_column-bzone;
+
+    if( sel_column+bzone >= first_column + row_width )
+        first_column=sel_column-row_width+bzone+1;
+
+   if(first_column<0)
+       first_column=0;
+
+    if(first_column>=NUM_LABORS-row_width)
+        first_column=NUM_LABORS-row_width;
 
-    // if it shrinks horizontally, scroll to the right to keep the cursor visible
-    if (first_column < sel_column - col_widths[DISP_COLUMN_LABORS] + 1)
-        first_column = sel_column - col_widths[DISP_COLUMN_LABORS] + 1;
+}
+
+bool viewscreen_unitlaborsst::scrollknock(int *reg, int stickval, int passval){
```

Comment:
Is this supposed to wrap around after a certain time or something?

Reply:
It requires a double tap to wrap around. This creates ergonomically comfortable edges, accidental and unexpected jumping is disorientating in user experience. Measuring a system time was necessary to implement this behaviour reliably.